modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/OptimizingPhaseHandler.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.ignite.internal.sql.engine.sql.ParsedResult;
+import org.apache.ignite.internal.sql.engine.tx.QueryTransactionContext;
+import org.apache.ignite.internal.sql.engine.tx.QueryTransactionWrapper;
+import org.apache.ignite.internal.sql.engine.util.TypeUtils;
+import org.apache.ignite.sql.SqlException;
+
+/** Validates parsed AST acquired on the previous phase and submit optimization task to {@link PrepareService}. */
+class OptimizingPhaseHandler implements ExecutionPhaseHandler {
+    static final ExecutionPhaseHandler INSTANCE = new OptimizingPhaseHandler();
+
+    private OptimizingPhaseHandler() {
+    }
+
+    @Override
+    public Result handle(Query query) {
+        ParsedResult result = query.parsedResult;
+
+        assert result != null : "Query is expected to be parsed at this phase";
+
+        validateParsedStatement(query.properties, result);
+        validateDynamicParameters(result.dynamicParamsCount(), query.params, true);
+        ensureStatementMatchesTx(result.queryType(), query.txContext);
+
+        HybridTimestamp operationTime = query.executor.deriveOperationTime(query.txContext);
+
+        String schemaName = query.properties.get(QueryProperty.DEFAULT_SCHEMA);
+        ZoneId timeZoneId = query.properties.get(QueryProperty.TIME_ZONE_ID);
+
+        SqlOperationContext operationContext = SqlOperationContext.builder()
+                .queryId(query.id)
+                .cancel(query.cancel)
+                .prefetchCallback(new PrefetchCallback())
+                .parameters(query.params)
+                .timeZoneId(timeZoneId)
+                .defaultSchemaName(schemaName)
+                .operationTime(operationTime)
+                .txContext(query.txContext)
+                .txUsedListener(tx -> query.usedTransaction = tx)
+                .build();
+
+        query.operationContext = operationContext;
+
+        CompletableFuture<Void> awaitFuture = query.executor.waitForMetadata(operationTime)
+                .thenCompose(none -> query.executor.prepare(result, operationContext)
+                        .thenAccept(plan -> {
+                            if (query.txContext.explicitTx() == null) {
+                                // in case of implicit tx we have to update observable time to prevent tx manager to start
+                                // implicit transaction too much in the past where version of catalog we used to prepare the
+                                // plan was not yet available
+                                query.txContext.updateObservableTime(query.executor.deriveMinimalRequiredTime(plan));
+                            }
+
+                            query.plan = plan;
+
+                            query.moveTo(ExecutionPhase.CURSOR_INITIALIZATION);
+                        }));
+
+        return Result.proceedAfter(awaitFuture);
+    }
+
+    /** Performs additional validation of a parsed statement. **/
+    private static void validateParsedStatement(
```

Comment:
Looks like this function duplicates the `validateParsedStatement` from the `SqlQueryProcessor`. Perhaps it would be better to define this function once.

Reply:
good catch! Moved all such validation to `ValidationHelper` in order to reuse it